module/modmanager/manager.go

```diff
@@ -0,0 +1,300 @@
+// Package modmanager provides the module manager for a robot.
+package modmanager
+
+import (
+	"context"
+	"io/fs"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+	pb "go.viam.com/api/module/v1"
+	"go.viam.com/utils/pexec"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"go.viam.com/rdk/config"
+	rdkgrpc "go.viam.com/rdk/grpc"
+	modlib "go.viam.com/rdk/module"
+	modif "go.viam.com/rdk/module/modmanager/modmaninterface"
+	"go.viam.com/rdk/operation"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot"
+)
+
+// NewManager returns a Manager.
+func NewManager(r robot.LocalRobot) (modif.ModuleManager, error) {
+	return &Manager{
+		logger:  r.Logger().Named("modmanager"),
+		modules: map[string]*module{},
+		r:       r,
+		rMap:    map[resource.Name]*module{},
+	}, nil
+}
+
+type module struct {
+	name    string
+	process pexec.ManagedProcess
+	handles modlib.HandlerMap
+	conn    *grpc.ClientConn
+	client  pb.ModuleServiceClient
+	addr    string
+}
+
+// Manager is the root structure for the module system.
+type Manager struct {
+	mu      sync.RWMutex
+	logger  golog.Logger
+	modules map[string]*module
+	r       robot.LocalRobot
+	rMap    map[resource.Name]*module
+}
+
+// Close terminates module connections and processes.
+func (mgr *Manager) Close(ctx context.Context) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	var err error
+	for _, mod := range mgr.modules {
+		if mod.conn != nil {
+			err = multierr.Combine(err, mod.conn.Close())
+		}
+		if mod.process != nil {
+			err = multierr.Combine(err, mod.process.Stop())
+		}
+	}
+	return err
+}
+
+// AddModule adds and starts a new resource module.
+func (mgr *Manager) AddModule(ctx context.Context, cfg config.Module) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	_, exists := mgr.modules[cfg.Name]
+	if exists {
+		return nil
+	}
+
+	mod := &module{}
+	mgr.modules[cfg.Name] = mod
+
+	parentAddr, err := mgr.r.ModuleAddress()
+	if err != nil {
+		return err
+	}
+	mod.addr = filepath.Dir(parentAddr) + "/" + cfg.Name + ".sock"
+
+	pcfg := pexec.ProcessConfig{
+		ID:   cfg.Name,
+		Name: cfg.ExePath,
+		Args: []string{mgr.modules[cfg.Name].addr},
+		Log:  true,
+	}
+	mod.process = pexec.NewManagedProcess(pcfg, mgr.logger)
+
+	err = mod.process.Start(context.Background())
+	if err != nil {
+		return errors.WithMessage(err, "module startup failed")
+	}
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	for {
```

Comment:
might want to refactor this function into a few smaller functions to make testing and reading easier